### PR TITLE
Added before_fetch_job hook

### PR DESCRIPTION
**Pull request by gena-admin**
https://coupadev.atlassian.net/browse/CD-336575

- [ ] @hubenias

## Generated Reviewers

<details>
  <summary><em>:warning: Do not modify anything in this section! :warning:</em></summary>
  <em>The content in this section is managed automatically, and any manual changes to it will be ignored and overwritten the next time the code review status is refreshed! View <a href="https://cody.coupa.engineering/repos/coupa/resque/pull/12">this PR</a> on Cody to see the current code review status.</em>
</details>

### CODE REVIEW

File: lib/resque.rb [hunk 1]
```diff
@@ -189,6 +189,11 @@ def send_queues_stats(&block)
     block ? register_hook(:send_queues_stats, block) : hooks(:send_queues_stats)
   end
 
+  # If the hook returns false, the method fetch_job will not be call.
+  def before_fetch_job(&block)
```
Comment (johnny-lai) [inline lib/resque.rb:193]: Minor: We should document the parameters that are passed to the hook as well.

File: lib/resque/worker.rb [hunk 2]
```diff
@@ -334,6 +334,12 @@ def reserve
     # Given a string queue name, returns an instance of Resque::Job
     # if any jobs are available. If not, returns nil.
     def fetch_job(queue)
+      result = run_hook :before_fetch_job, queue, redis do |results|
+        !results.any? { |result| result == false }
+      end
```
Comment (johnny-lai) [inline lib/resque/worker.rb:339]: So I see `run_hook` returns a result and the hooks depending on different  parameters. That feels a little unexpected.

```
event = OpenStruct.new(cancel: false)
run_hook :before_fetch_job, event, queue, redis
return if event.cancel == true
```

This style is similar to the one use for `before_term_signal`

Comment (gena-admin) [reply lib/resque/worker.rb:339]: 👍

Comment (gena-admin) [reply lib/resque/worker.rb]: Done

File: lib/resque.rb (hunk 1 above)
Comment (gena-admin) [reply lib/resque.rb:194]: Done

**Review by johnny-lai — APPROVED** [review]

File: lib/resque/worker.rb [hunk 3]
```diff
@@ -334,6 +334,10 @@ def reserve
     # Given a string queue name, returns an instance of Resque::Job
     # if any jobs are available. If not, returns nil.
     def fetch_job(queue)
+      event = OpenStruct.new(cancel: false)
+      run_hook :before_fetch_job, event, queue, redis
+      return if event.cancel == true
```
Comment (johnny-lai) [inline lib/resque/worker.rb:339]: Maybe 
`return nil if event.cancel == true` to be clearer.

**Review by hubenias — APPROVED** [review]

**Pull request merged**